ext/storage_sources/s3_store/aws_bucket_conn.cpp

```diff
@@ -0,0 +1,119 @@
+#include <aws/core/Aws.h>
+#include <aws/s3-crt/S3CrtClient.h>
+#include <aws/s3-crt/model/DeleteObjectRequest.h>
+#include <aws/s3-crt/model/ListObjectsRequest.h>
+#include <aws/s3-crt/model/Object.h>
+#include <aws/s3-crt/model/PutObjectRequest.h>
+#include "aws_bucket_conn.h"
+
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+/*
+ * list_buckets --
+ *     Return list of buckets from AWS account.
+ */
+bool
+aws_bucket_conn::list_buckets(std::vector<std::string> &buckets) const
+{
+    Aws::S3Crt::Model::ListBucketsOutcome outcome = m_s3_crt_client.ListBuckets();
+    if (outcome.IsSuccess()) {
+        for (Aws::S3Crt::Model::Bucket bucket : outcome.GetResult().GetBuckets())
+            buckets.push_back(bucket.GetName());
+        return true;
+    } else {
+        std::cerr << "Error in list_buckets: " << outcome.GetError().GetMessage() << std::endl
+                  << std::endl;
+        return false;
+    }
+}
+
+/*
+ * list_objects --
+ *     Return list of objects from S3 bucket.
+ */
+bool
+aws_bucket_conn::list_objects(
+  const Aws::String bucket_name, std::vector<Aws::S3Crt::Model::Object> &bucket_objects) const
+{
+    Aws::S3Crt::Model::ListObjectsRequest request;
+    request.WithBucket(bucket_name);
+    Aws::S3Crt::Model::ListObjectsOutcome outcomes = m_s3_crt_client.ListObjects(request);
+
+    if (outcomes.IsSuccess()) {
+        bucket_objects = outcomes.GetResult().GetContents();
+        return true;
+    } else {
+        std::cerr << "Error in list_buckets: " << outcomes.GetError().GetMessage()
+                  << std::endl;
+        return false;
+    }
+}
+
+/*
+ * put_object --
+ *     Put an object into S3 bucket.
+ */
+bool
+aws_bucket_conn::put_object(
+  const Aws::String &bucket_name, const Aws::String &object_key, const Aws::String &file_name) const
+{
+    Aws::S3Crt::Model::PutObjectRequest request;
+    request.SetBucket(bucket_name);
+    request.SetKey(object_key);
+
+    std::shared_ptr<Aws::IOStream> input_data = Aws::MakeShared<Aws::FStream>(
```

Comment:
Why does it need to be a `shared_ptr`? Can it be a `unique_ptr`?

Reply:
Seems that this is what they use in their [docs](https://docs.aws.amazon.com/pt_br/sdk-for-cpp/v1/developer-guide/examples-s3-objects.html).